### DISCUSSION TITLE
Terminal module: Add an option to automatically switch to user's working directory

### DISCUSSION THREAD

**Discussion by iliaross** [webmin/webmin#2204] [General]
Hey Jamie,

On some development systems, multiple users are allowed to log in as _root_, which can lead to permission issues for files created under users' home directories, as was pointed out in https://github.com/virtualmin/virtualmin-gpl/issues/836.

To automate the prevention of this, I suggest adding an option to the Terminal module that will allow switching to the actual user's home when logged in as _root_. During testing, I simply added the following to the `/root/.bashrc` file (the actual implementation in the Terminal module might differ):

```bash
switch_user_on_cd() {
  # Extract the username from the path if it's under /home
  if [[ "$PWD" == /home/* ]]; then
    username=$(basename "$PWD")
    
    # Check if the extracted username exists on
    # the system and switch automatically
    if id -u "$username" > /dev/null 2>&1; then
      su - "$username"
    fi
  fi
}

# Set the PROMPT_COMMAND to call the switch_user_on_cd function
PROMPT_COMMAND=switch_user_on_cd
```

By setting the `PROMPT_COMMAND` environment variable, our custom `switch_user_on_cd` function is executed before each prompt display. The _root_ user can then type `exit` or hit `Ctrl+D` to revert to _root_. This feature already exists in the File Manager, where permissions are automatically handled when working as _root_ in users' home directories.

**Comment by iliaross** [comment]
I've actually got excited and came up with a complete working solution now!

Here's a reworked variant that allows the _root_ user to automatically switch to the home user and back without any extra effort. By setting `ROOT=1`, it's possible to force the use of _root_ in the user's home directory.

All that needs to be done is to create a file named `switch_user_on_cd.sh` under the `/etc/profile.d` directory with the following content:

```bash
# Extract username from the home directory
get_username_from_home() {
  local home_dir="$1"
  local username=$(awk -F: -v home="$home_dir" '$6 == home {print $1}' /etc/passwd)
  echo "$username"
}

# Function to switch user on cd
switch_user_on_cd() {
  # If root is forced under for whatever reason exiit
  if [ -n "$ROOT" ]; then
    return
  fi

  # Check if the current directory is not under /home
  if [[ "$PWD" != /home/* ]]; then
    if [[ -n "$ORIGINAL_USER" && "$(whoami)" != "root" ]]; then
      exit
    fi
  else
    # Extract the username from the home directory if it's under /home
    username=$(get_username_from_home "$PWD")
    if [[ -n "$username" ]]; then
      if [[ "$(whoami)" != "$username" ]]; then
        export ORIGINAL_USER=$(whoami)
        export ORIGINAL_USER_PWD=$PWD
        export ORIGINAL_USER_OLDPWD=$OLDPWD
        sudo -H -u "$username" bash -c 'export ORIGINAL_USER='"$ORIGINAL_USER"'; export ORIGINAL_USER_PWD='"$ORIGINAL_USER_PWD"'; export ORIGINAL_USER_OLDPWD='"$ORIGINAL_USER_OLDPWD"';  exec bash -l'
        if [ "$PWD" = "$ORIGINAL_USER_PWD" ]; then
          cd "$ORIGINAL_USER_OLDPWD"
        fi
        unset ORIGINAL_USER
        unset ORIGINAL_USER_PWD
        unset ORIGINAL_USER_OLDPWD
      fi
    fi
  fi
}

# Set the PROMPT_COMMAND to call the switch_user_on_cd function
PROMPT_COMMAND=switch_user_on_cd
```

Give it a try. Do you feel the magic? 🙂

**Comment by jcameron** [comment]
That looks like a complex solution to the problem!

What kind of user is able to login as root but has a different home directory?

**Reply by iliaross** [reply]
> What kind of user is able to login as root but has a different home directory?

Umm .. I don't understand your question.

Using Virtualmin terminology, this code _automatically_ switches to the domain owner when the master administrator visits the domain home directory.

In other words, if logged in as _root_ and navigating to `/home/example1.com`, all files created under the user home will be owned by the `example1` user instead of _root_. When going outside of the home directory, it will switch back to _root_. 🙂 

Please give it a try and let me know what you think?

**Reply by jcameron** [reply]
What I mean is, what's the user problem that this patch is trying to fix?

Is it that when logging in a `root` and changing to a domain's home directory that files are created with `root` permissions and not those of the domain owner?

**Reply by iliaross** [reply]
> Is it that when logging in a `root` and changing to a domain's home directory that files are created with `root` permissions and not those of the domain owner?

Yeah, right.

> What I mean is, what's the user problem that this patch is trying to fix?
> 
> Is it that when logging in a `root` and changing to a domain's home directory that files are created with `root` permissions and not those of the domain owner?

This isn't a patch yet, just a discussion.

**Reply by jcameron** [reply]
Ok, got it it! However, I really don't think we want to add this kind of complexity to user's .bashrc files (even as an option) to deal with a situation that could be resolved by just su-ing to the correct user.

**Reply by iliaross** [reply]
We already have `xterm/rc/.bashrc` .. Besides it cannot be done  using `.bashrc` file alone. It should be done using `/etc/profile.d` .. 

> to deal with a situation that could be resolved by just su-ing to the correct user.

Does every beginner user know about this?

**Reply by iliaross** [reply]
> It should be done using `/etc/profile.d` ..

I don't even think it should be a Terminal module option ..

**Reply by jcameron** [reply]
I mean I'm ok if users want to add this and we could even recommend it in docs, but I don't think it's something we'd want to setup automatically..

**Reply by iliaross** [reply]
Perhaps we could add a tick under **Webmin ⇾ Webmin Configuration: Operating System and Environment** page somewhere?

**Reply by jcameron** [reply]
I'm not even sure if we want that, because this change leads to some surprising behavior. Like what if the `root` user actually wants to edit files in a domain owner's directory with `root` privileges?

**Reply by iliaross** [reply]
Well, this is almost never should be needed though it's possible to do currently by setting `ROOT=1`  env var.

**Reply by jcameron** [reply]
I mean this is clever, but I don't thunk we should ever offer it as an option in Virtualmin because changing what the "cd" command does goes against common practice and what users expect.

**Reply by iliaross** [reply]
I won't disagree. However, I think, that some people could find it useful.

But okay, let's then just document it in Virtualmin maybe at least?

**Reply by jcameron** [reply]
Oh yeah documenting this would be reasonable..

**Comment by iliaross** [comment]
Alright, documented here:

https://www.virtualmin.com/docs/advanced-tutorials/auto-user-switch-in-shell-navigation
